odoo/addons/base/models/res_partner.py

```diff
@@ -827,66 +827,10 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
-        self.recompute(['display_name'])
-        self.flush()
-        if args is None:
-            args = []
-        order_by_rank = self.env.context.get('res_partner_search_mode')
-        if (name or order_by_rank) and operator in ('=', 'ilike', '=ilike', 'like', '=like'):
-            self.check_access_rights('read')
-            where_query = self._where_calc(args)
-            self._apply_ir_rules(where_query, 'read')
-            from_clause, where_clause, where_clause_params = where_query.get_sql()
-            from_str = from_clause if from_clause else 'res_partner'
-            where_str = where_clause and (" WHERE %s AND " % where_clause) or ' WHERE '
-
-            # search on the name of the contacts and of its company
-            search_name = name
-            if operator in ('ilike', 'like'):
-                search_name = '%%%s%%' % name
-            if operator in ('=ilike', '=like'):
-                operator = operator[1:]
-
-            unaccent = get_unaccent_wrapper(self.env.cr)
-
-            fields = self._get_name_search_order_by_fields()
-
-            query = """SELECT res_partner.id
-                         FROM {from_str}
-                      {where} ({email} {operator} {percent}
-                           OR {display_name} {operator} {percent}
-                           OR {reference} {operator} {percent}
-                           OR {vat} {operator} {percent})
-                           -- don't panic, trust postgres bitmap
-                     ORDER BY {fields} {display_name} {operator} {percent} desc,
-                              {display_name}
-                    """.format(from_str=from_str,
-                               fields=fields,
-                               where=where_str,
-                               operator=operator,
-                               email=unaccent('res_partner.email'),
-                               display_name=unaccent('res_partner.display_name'),
-                               reference=unaccent('res_partner.ref'),
-                               percent=unaccent('%s'),
-                               vat=unaccent('res_partner.vat'),)
-
-            where_clause_params += [search_name]*3  # for email / display_name, reference
-            where_clause_params += [re.sub(r'[^a-zA-Z0-9\-\.]+', '', search_name) or None]  # for vat
-            where_clause_params += [search_name]  # for order by
-            if limit:
-                query += ' limit %s'
-                where_clause_params.append(limit)
-            self.env.cr.execute(query, where_clause_params)
-            return [row[0] for row in self.env.cr.fetchall()]
-
-        return super(Partner, self)._name_search(name, args, operator=operator, limit=limit, name_get_uid=name_get_uid)
+        # TODO vat must be sanitized the same way for storing/searching
+        return self._name_search_fields(name, ['email', 'display_name', 'ref', 'vat'], args, operator, limit, name_get_uid)
```

Comment:
I keep the order when the context is set with `res_partner_search_mode`, but I indeed change the order in other case like you said.
I feel like the way I implement the `_order` in `addons/account/models/partner.py` is an improvement.
That order really seems like arbitrary to me though. Wouldn't a sort done in python be acceptable if it is really needed? I find this code a bit too complicated.

Note that I'm actually fixing the broken search on the VAT here by removing any "smartness" to it

Reply:
The order in the SQL query is not arbitrary, it is what a user expects.  If you search with name "foo", you want the contacts where the name contains "foo" to appear first.  Then will appear the ones matching "foo" in their `ref` or `vat`.  This makes a big difference when search has a _limit_, which is the most common case, because _less relevant matches_ will probably _not appear at all within the limit_.

Reply:
> it is what a user expects

In some cases.
If I type "cust", I might want all the partners with a `ref` containing `cust` first. If I type "no", I might want the partners with a Norwegian vat number first. If I type "gmail" (or "the-name-of-my-company"), I might want all the users with a Gmail account (or from my company).
So I stand my point that it is making the code way more complex for something that is arbitrary.
I any case, if you want something that is not arbitrary, you have the search view. And if you want to filter out things that don't interest you, be more specific.

Otherwise, you might also want to say that if something matches exactly what you inputted, you want it to be the top of the list, exactly like for countries for their codes.

Reply:
From the answers I got on the PO's channel, nobody seems to care. [The only answer I got](https://discord.com/channels/678381219515465750/687337648775495699/956638780700373062) was even in favor of this change.